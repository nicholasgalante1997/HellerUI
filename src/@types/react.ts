import React from 'react';

export type SafeReactFC<T> = T & {
    children?: React.ReactNode | React.ReactNode[];
};
