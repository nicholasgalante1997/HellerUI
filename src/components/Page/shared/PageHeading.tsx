import React from 'react';
import * as Font from '../../Typography';
import { BasePage as BasePageType } from '../types';
import ActionBar from './ActionBar';
import { Break } from '../../Break';
import { HellerDivider } from '../../Divider';
import { getSubtitleFontStyles } from '../utils';

const { Paragraph, Heading } = Font;
const TitleSection = Heading;
const SubTitle = Paragraph;

type PageHeadingProps = Pick<
  BasePageType<{}>,
  | 'withActionBar'
  | 'withDividers'
  | 'dividerProps'
  | 'dangerouslyOverridePageHeading'
  | 'title'
  | 'titleColor'
  | 'subtitle'
  | 'subtitleColor'
  | 'subtitleSize'
  | 'titleSize'
>;

const PageHeading = ({
  withActionBar,
  withDividers,
  dividerProps,
  dangerouslyOverridePageHeading,
  title,
  titleColor,
  subtitle,
  subtitleColor,
  subtitleSize,
  titleSize
}: PageHeadingProps) => {

  if (dangerouslyOverridePageHeading?.headingNode) {
    return dangerouslyOverridePageHeading.headingNode
  }

  return (
  <>
    {/* Action Bar */}
    {withActionBar ? (
      <>
        <ActionBar {...withActionBar} />
        <Break />
      </>
    ) : null}

    {/* Page Title */}
    <TitleSection as={titleSize} scale={500} color={titleColor}>
      {title}
    </TitleSection>
    <Break height="0.5rem" />

    {/* Section Subtitle */}
    {subtitle ? (
      <SubTitle
        customStyles={{
          lineHeight: getSubtitleFontStyles(subtitleSize!).lineHeight,
          fontWeight: getSubtitleFontStyles(subtitleSize!).fontWeight
        }}
        thin
        color={subtitleColor}
        fontSize={getSubtitleFontStyles(subtitleSize!).fontSize}
      >
        {subtitle}
      </SubTitle>
    ) : null}

    {/* Section Title Break */}
    {withDividers ? (
      <HellerDivider {...dividerProps!} data-testid="hr" />
    ) : (
      <Break height="1.5rem" />
    )}
  </>
)
    };

export default PageHeading;
