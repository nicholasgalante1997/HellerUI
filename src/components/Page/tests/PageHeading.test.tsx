/* eslint-disable no-restricted-syntax */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-undef */
import React from 'react';
import { render } from '@testing-library/react';
import PageHeading from '../shared/PageHeading';

describe('Page Heading Component', () => {
  test('PageHeading Default', () => {
    const { container, queryAllByText, queryByTestId } = render(
      <PageHeading
        subtitleColor="red"
        titleColor="black"
        title="heller ui"
        subtitle="test subtitle"
        withDividers
        dividerProps={{ fadeColor: 'yellow', focusColor: 'pink' }}
      />
    );

    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toBeInTheDocument();

    const h = queryAllByText('heller ui');
    const sub = queryAllByText('test subtitle');

    expect(sub[0]).toHaveStyle("font-family: 'Poppins Thin';");
    expect(sub[0]).toHaveStyle('font-size: 32px;');
    expect(sub[0]).toHaveStyle('color: red;');

    expect(h[0]).toHaveStyle("font-family: 'Poppins Bold';");
    expect(h[0]).toHaveStyle('color: black;');

    // rogue element capture
    const breakEl = queryByTestId('br');
    expect(breakEl).not.toBeNull();

    const dividerEl = queryByTestId('hr');
    expect(dividerEl).not.toBeNull();
  });
  test('PageHeading ActionBar', () => {
    const { queryByTestId } = render(
      <PageHeading
        subtitleColor="red"
        title="heller ui"
        subtitle="test subtitle"
        withActionBar={{
          actionTitle: <h1 data-testid="ph-action-bar">mock</h1>
        }}
        withDividers
        dividerProps={{ fadeColor: 'yellow', focusColor: 'pink' }}
      />
    );

    const el = queryByTestId('ph-action-bar');
    expect(el).not.toBeNull();
    expect(el).toBeInTheDocument();
  });
  test('title as states', () => {
    const sizes = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;
    for (const s of sizes) {
      const { unmount } = render(<PageHeading title="mock" titleSize={s} />);
      const heading = document.querySelector(s);
      expect(heading).not.toBeNull();
      expect(heading).toBeInTheDocument();
      unmount();
    }
  });
  test('PageHeading no subtitle', () => {
    render(
      <PageHeading
        title="heller ui"
        withDividers
        dividerProps={{ fadeColor: 'yellow', focusColor: 'pink' }}
      />
    );
    const arr = Array.from(document.querySelectorAll('p'));
    expect(arr.length).toBe(0);
  });
  test('PageHeading no hrs', () => {
    render(<PageHeading title="heller ui" />);
    const arr = Array.from(document.querySelectorAll('hr'));
    expect(arr.length).toBe(0);
  });
});
