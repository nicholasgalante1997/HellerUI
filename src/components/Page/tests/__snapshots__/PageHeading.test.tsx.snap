// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Heading Component PageHeading override 1`] = `
<div
  data-testid="override"
>
  ov
</div>
`;
